src/kopls.js

```diff
@@ -0,0 +1,248 @@
+
+        var KX = this.kernel.compute(trainingSet);
+
+        var I = Matrix.eye(KX.rows, KX.rows, 1);
+        var Kmc = KX.clone();
+        KX = new Matrix(this.orthogonalComp + 1, this.orthogonalComp + 1);
+        KX[0][0] = Kmc;
+
+        var result = new SingularValueDecomposition(trainingValues.transpose().mmul(KX[0][0]).mmul(trainingValues), {
+            computeLeftSingularVectors: true,
+            computeRightSingularVectors: false
+        });
+        var Cp = result.leftSingularVectors;
+        var Sp = result.diagonalMatrix;
+
+        Cp = Cp.subMatrix(0, Cp.rows - 1, 0, this.predictiveComp - 1);
+        Sp = Sp.subMatrix(0, this.predictiveComp - 1, 0, this.predictiveComp - 1);
+
+        var Up = trainingValues.mmul(Cp);
+
+        var Tp = new Array(this.orthogonalComp + 1);
+        var Bt = new Array(this.orthogonalComp + 1);
+        var to = new Array(this.orthogonalComp);
+        var co = new Array(this.orthogonalComp);
+        var so = new Array(this.orthogonalComp);
+        var toNorm = new Array(this.orthogonalComp);
+
+        var SpPow = Matrix.pow(Sp, -0.5);
+        // to avoid errors, check infinity
+        SpPow.apply(function (i, j) {
+            if (this[i][j] === Infinity) {
+                this[i][j] = 0.0;
+            }
+        });
+
+        for (var i = 0; i < this.orthogonalComp; ++i) {
+            Tp[i] = KX[0][i].transpose().mmul(Up).mmul(SpPow);
+
+            var TpiPrime = Tp[i].transpose();
+            Bt[i] = inverse(TpiPrime.mmul(Tp[i])).mmul(TpiPrime).mmul(Up);
+
+            result = new SingularValueDecomposition(TpiPrime.mmul(Matrix.sub(KX[i][i], Tp[i].mmul(TpiPrime))).mmul(Tp[i]), {
+                computeLeftSingularVectors: true,
+                computeRightSingularVectors: false
+            });
+            var CoTemp = result.leftSingularVectors;
+            var SoTemp = result.diagonalMatrix;
+
+            co[i] = CoTemp.subMatrix(0, CoTemp.rows - 1, 0, 0);
+            so[i] = SoTemp[0][0];
+
+            to[i] = Matrix.sub(KX[i][i], Tp[i].mmul(TpiPrime)).mmul(Tp[i]).mmul(co[i]).mul(Math.pow(so[i], -0.5));
+
+            var toiPrime = to[i].transpose();
+            toNorm[i] = Matrix.sqrt(toiPrime.mmul(to[i]));
+
+            to[i] = to[i].divRowVector(toNorm[i]);
+
+            var ITo = Matrix.sub(I, to[i].mmul(to[i].transpose()));
+
+            KX[0][i + 1] = KX[0][i].mmul(ITo);
+            KX[i + 1][i + 1] = ITo.mmul(KX[i][i]).mmul(ITo);
+        }
+
+        var lastTp = Tp[this.orthogonalComp] = KX[0][this.orthogonalComp].transpose().mmul(Up).mmul(SpPow);
+
+        var lastTpPrime = lastTp.transpose();
+        Bt[this.orthogonalComp] = inverse(lastTpPrime.mmul(lastTp)).mmul(lastTpPrime).mmul(Up);
+
+        this.Cp = Cp;
+        this.SpPow = SpPow;
+        this.Up = Up;
+        this.Tp = Tp;
+        this.co = co;
+        this.so = so;
+        this.to = to;
+        this.toNorm = toNorm;
+        this.Bt = Bt;
+        this.K = KX;
```

Comment:
Is very difficult to follow the code if the variable names are like co, so, Tp, etc. Could you please change them to be a little bit more readable?